flink-core/src/main/java/org/apache/flink/api/common/TaskInfo.java

```diff
@@ -132,11 +122,67 @@ public String getTaskNameWithSubtasks() {
 		return this.taskNameWithSubtasks;
 	}
 
+	/**
+	 * Returns the name of the task, appended with the subtask indicator and execution attempt id.
+	 *
+	 * @return The name of the task, with subtask indicator and execution id.
+	 */
+	public String getTaskNameWithSubtaskAndID() {
+		return taskNameWithSubtaskAndID;
+	}
+
 	/**
 	 * Returns the allocation id for where this task is executed.
 	 * @return the allocation id for where this task is executed.
 	 */
 	public String getAllocationIDAsString() {
 		return allocationIDAsString;
 	}
+
+	// ------------------------------------------------------------------------
+	//  Factory Utilities
+	// ------------------------------------------------------------------------
+
+	public static TaskInfo create(
+		String taskName,
+		int maxNumberOfParallelSubtasks,
+		int numberOfParallelSubtasks,
+		int subtaskIndex,
+		int attemptNumber,
+		String attemptIDAsString,
```

Comment:
why passing a `String` and not a `ExecutionAttemptID`?

Reply:
Andrey gave me the answer that `ExecutionAttemptID` is not part of `flink-core`. But then I'm wondering, whether `TaskInfo` is the right container to transport the `ExecutionAttemptID` to the `BarrierBuffer`. It feels a bit that we are abusing a class which was not intended for this purpose here.

Reply:
Yes, it is because of module dependency.
I do not think we are abusing a class, because the previous usage of `allocationIDAsString` also has the same issue.
In the first version I generates the full task name with attempt id in task class and then pass this string full task name into constructor of `TaskInfo` which is not aware of how this name is generated. But Andrey suggests passing the original components as now to make task class simple logic.

Reply:
I think the `allocationIDAsString` is also not right and suffers from the same problems.

Reply:
I am not sure the history consideration of introducing `allocationIDAsString` here. If we think it is not right thing to do, then I could avoid following this way. 

In addition, do you think I should refactor the way of `allocationIDAsString` in a hotfix commit?